### PR TITLE
Support soma.X.shape(), initial implementation

### PR DESCRIPTION
**Pull request by johnkerl**
Context: #125 

Example:

```
import tiledbsc

soma = tiledbsc.SOMA('tiledb-data/pbmc3k_processed')

print(soma.X.data.shape())
print(soma.obs.shape())
print(soma.var.shape())
print(soma.obsm['X_pca'].shape())
print(soma.varm['PCs'].shape())
print(soma.obsp['distances'].shape())
print(soma.raw.X.data.shape())
print(soma.raw.var.shape())
```

```
(2638, 1838)
(2638, 4)
(1838, 1)
(2638, 50)
(1838, 50)
(2638, 2638)
(2638, 13714)
(13714, 1)
```

### CODE REVIEW

File: apis/python/src/tiledbsc/annotation_matrix.py [hunk 1]
```diff
@@ -30,6 +30,22 @@ def __init__(
...
+        Returns a tuple with the number of rows and number of columns of the `AnnotationMatrix`.
+        The row-count is the number of obs_ids (for `obsm` elements) or the number of var_ids (for
+        `varm` elements).  The column-count is the number of columns/attributes in the dataframe.
+        """
+        with self._open() as A:
+            # These TileDB arrays are string-dimensioned sparse arrays so there is no '.shape'.
+            # Instead we compute it ourselves.  See also:
+            # * https://github.com/single-cell-data/TileDB-SingleCell/issues/10
+            # * https://github.com/TileDB-Inc/TileDB-Py/pull/1055
+            num_rows = len(A[:][self.dim_name].tolist())
```
Comment (ebezzi) [inline apis/python/src/tiledbsc/annotation_matrix.py:45]: We could replace `tolist()` with something like `sum(1 for _ in iterable)` and that won't fetch the whole list in memory.

Also, users would likely expect a method like `.shape` to be `O(1)` while this is effectively `O(n)`. I would perhaps add a note to the documentation to specify this.

Comment (johnkerl) [reply apis/python/src/tiledbsc/annotation_matrix.py:45]: good call @ebezzi ; discussing with @Shelnutt2 now as well ...

Comment (johnkerl) [reply apis/python/src/tiledbsc/annotation_matrix.py:47]: * Updated doc-blocks
* Added a tracking note to https://github.com/single-cell-data/TileDB-SingleCell/issues/125

Comment (johnkerl) [reply apis/python/src/tiledbsc/annotation_matrix.py:47]: P.S. @ebezzi I tried`tolist()` as well as `sum` on a 7GB dataset and found no meaningful difference:

```
import tiledbsc
import time

soma = tiledbsc.SOMA('/Users/johnkerl/mini-corpus/tiledb-data/azimuth-meta-analysis')

for k in range(5):
    t1 = time.time()
    print(soma.obs.shape())
    t2 = time.time()
    print("soma.obs.shape() seconds:", t2-t1)

    t1 = time.time()
    with soma.obs._open() as A:
        print(len(A[:]['obs_id'].tolist()))
    t2 = time.time()
    print("with-open-len() seconds:", t2-t1)

    t1 = time.time()
    with soma.obs._open() as A:
        print(sum(1 for _ in A[:]['obs_id']))
    t2 = time.time()
    print("with-open-sum() seconds:", t2-t1)
```

Sorting the outputs and taking average elapsed seconds I found:

```
# (982538, 27) soma.obs.shape() seconds: 14.805514097213745
# (982538, 27) soma.obs.shape() seconds: 4.8801109790802
# (982538, 27) soma.obs.shape() seconds: 13.00051474571228
# (982538, 27) soma.obs.shape() seconds: 6.960161209106445
# (982538, 27) soma.obs.shape() seconds: 12.627025842666626
#
# 10.4546653747559
#
# 982538 with-open-len() seconds: 12.833094120025635
# 982538 with-open-len() seconds: 11.15531587600708
# 982538 with-open-len() seconds: 12.893481016159058
# 982538 with-open-len() seconds: 4.761070251464844
# 982538 with-open-len() seconds: 13.486767053604126
#
# 11.0259456634521
#
# 982538 with-open-sum() seconds: 5.2499377727508545
# 982538 with-open-sum() seconds: 12.795608758926392
# 982538 with-open-sum() seconds: 12.787294864654541
# 982538 with-open-sum() seconds: 7.818833827972412
# 982538 with-open-sum() seconds: 13.448123216629028
#
# 10.4199596881866
```

**Pull request merged**